### DISCUSSION TITLE
General Discussion

### DISCUSSION THREAD

**Discussion by ThePhilgrim** [ThePhilgrim/MantaTail#65] [General]
@Akuli 
I have now started a Github Discussion here.
We can use this for discussing code (ideas, features, bugs, debugging, etc.)

I'll of course be on IRC when I can as well :)

EDIT: Discussions here do not have to be about Mantatail. It can serve as a general discussion whatever topic!

**Comment by ThePhilgrim** [comment]
Copying your answer on my latest question here, so that we can keep PR comments on-topic:

"Continuing off-topic, a few possible debugging steps for nc'ing from another computer:

Try to ping the computer: ping 192.168.1.91. If the computers find each other with that IP, it should print a bunch of stuff; try ping localhost or ping google.com to see what it should look like. If nothing is printed, or you get error messages, make sure the computers are connected to the same network (e.g. wifi) and the IP is correct. It could also be that if it's hotel wifi or something, it blocks computers from connecting to each other.
If pinging worked, try passing -v (verbose) to netcat, as in nc -v 192.168.1.91 6667.
I don't know much about networking details though, so if those steps don't help, I don't know how to troubleshoot this further."

**Comment by ThePhilgrim** [comment]
@Akuli After discussing the connection thingy with somebody on `#networking`, they said the port s not properly listening as it should and the "daemon is not open".
Do you happen to know what this means?
Adding screenshot:

![Skärmavbild 2022-01-17 kl  16 20 29](https://user-images.githubusercontent.com/76887896/149796468-6ec90b1c-94fc-4d55-8c78-fb583aa32e57.png)

**Reply by Akuli** [reply]
The relevant part of the discussion is: `well, maybe the server isn't allowing external connections`

Compare your `.bind()` call to dabeaz's `.bind()`. Also look up some documentation about `.bind()`. Maybe you'll figure it out :)

Meanwhile, make sure that netcatting between the two computers works. If it doesn't, we have a second problem too.

**Reply by ThePhilgrim** [reply]
Hm, would you please remind me how to netcat between two terminals without any specific server? I remember us doing this at one point.

Is the `.bind()` thing from his socket video that you showed me?

**Reply by Akuli** [reply]
Yes, `.bind()` is from that video.

On my system, the netcat command for making a server is `nc -lvp 6667`, where `6667` is the port number. It was different on mac. Maybe the same without `p`?

**Reply by ThePhilgrim** [reply]
Now I ran `nc -lv 1234` on my computer and connected to my IP from my girlfriend's computer. That worked well with no issues. However, when mantatail is running, it doesn't work (`Connection refused`)

I will check the bind setting thingy. Any other ideas apart from that?

**Reply by Akuli** [reply]
If netcat works but mantatail doesn't, it's the bind thingy :) The `.bind()` call specifies who is allowed to connect to the server, and on what port.

**Comment by ThePhilgrim** [comment]
@Akuli This might be an interesting read for you:

> Character Encodings
> Character encodings in IRC are hard. UTF-8 is recommended, the mess of Latin-1/ISO-8859-1(5)/CP1252 also seems common, but all sorts of other encodings are also used in practice. Particularly on networks that support other languages, and were created before UTF-8 became as widespread as it has.
> 
> When sending, we always recommend UTF-8. When decoding, we generally recommend trying UTF-8 and falling back to Latin-1 (what has been called the Hybrid encoding).
> 
> For clients, this is fine. Even if they incorrectly decode a private message, the user should see that the message has been decoded incorrectly and be able to resolve the issue (hopefully by telling the sending user to use UTF-8).
> 
> However, servers are in a trickier position (especially for PRIVMSG/NOTICE or any other command that takes arbitrary user input such as USER, TOPIC, etc). Servers in C/C++ or similar languages can simply treat this input from the user as a character array they accept and then spit out again, no trouble. However, servers implemented in higher-level languages may wish to treat IRC lines and messages as Unicode text internally. For servers to treat messages in this way, they need to decode lines as they’re received and later encode the lines before they’re sent out.
> 
> This presents an issue. What if the line from the user is decoded incorrectly, and then this modified line is sent out? (see also: Mojibake). What these servers may instead do is either follow the lead of the lower-level servers and treat these parameters as byte arrays not to be parsed or decoded in any way. These servers may also simply attempt to decode all incoming lines as UTF-8, and if the line cannot be decoded it is ignored. The former ensures all messages are sent correctly, and the latter vastly simplifies server implementations in these higher-level languages.
> 
> For what it’s worth, the servers I’ve worked on have done the latter of the two options, discarding any lines that cannot be decoded as UTF-8. It’s not a great solution, but it’s a difficult issue to deal with. If anyone has a better way to deal with this, please feel free to contact me so I can update this section.

Found in the [IRC Horse Docs](https://modern.ircdocs.horse/#character-encodings)
You might actually want to contact this person.

**Reply by Akuli** [reply]
> Servers in C/C++ or similar languages can simply treat this input from the user as a character array they accept and then spit out again, no trouble. 

This is what we do, in a high level language, in exactly the same way it would work in C or C++. I'll add a reminder for myself to eventually contact whoever wrote this.

**Reply by ThePhilgrim** [reply]
@Akuli If you go to the link I put in the comment, there is a link to contact to person.

**Reply by Akuli** [reply]
I sent an email. I will let you know if they respond.

**Comment by Akuli** [comment]
I created a simple fuzzer for mantatail. If you have never heard of fuzzers, the beginning of wikipedia article explains the concept well: https://en.wikipedia.org/wiki/Fuzzing

```python3
import socket
import random


words = [
    "#bar",
    "#foo",
    "+g",
    "+o",
    "+xyz"
    "-g",
    "-o",
    "-xyz"
    "JOIN",
    "KICK",
    "MODE",
    "NICK",
    "PART",
    "PING",
    "PONG",
    "PRIVMSG",
    "QUIT",
    "USER",
    "",
]

while True:
    command = "".join(
        " ".join(
            random.choice(words)
            for word_number in range(random.randint(1, 5))
        ) + "\n"
        for line_number in range(100)
    )

    sock = socket.socket()
    sock.connect(('localhost', 6667))
    sock.sendall(command.encode())
    sock.shutdown(socket.SHUT_RDWR)
    sock.close()
```

To use it, run mantatail, and then run the fuzzer in another terminal. I think you may need to comment out the .shutdown() line on MacOS.

Mantatail shouldn't error, no matter what a client sends to it. But it errors a lot when I run the fuzzer. There's already issues for some of the bugs. When they are fixed, we should run the fuzzer again to see if more bugs remain.

The fuzzer creates random strings that look like IRC commands by joining together 1 to 5 words, where each word is taken from the `words` list. The result could look like `b"JOIN PRIVMSG #foo Alice\r\n"`, for example. We send 100 of these at a time to Mantatail.

**Comment by ThePhilgrim** [comment]
@Akuli I'm currently looking into the ban feature.

Usually, a moderator can set several modes in the same command. Examples:

```
MODE #foobar +mb *@127.0.0.1    ; Setting the "moderated" channel mode and
        adding the "*@127.0.0.1" mask to the ban
        list of the #foobar channel.
```
```
:dan!~h@localhost MODE #foobar -bl+i *@192.168.0.1
        ; dan unbanned the "*@192.168.0.1" mask,
        removed the client limit from, and set the
        #foobar channel to invite-only.
```

This doesn't seem too difficult to implement. However, if I log into Libera with two netcats, and with a channel operator I do `MODE #foo +ob Target_user`, it sets Target_user as operator, but doesn't ban Target_user (yes, you can ban an operator). 
Instead the command is interpreted as `MODE #foo +o Target_user` & `MODE #foo +b` (which sends the ban list to the user)

![Skärmavbild 2022-01-28 kl  22 43 04](https://user-images.githubusercontent.com/76887896/151626206-e86c5f14-8409-48c2-9f4f-47a28fbbb1ec.png)

I'm confused about why it doesn't connect both +o and +b to Target_user as it should.

This makes it much harder to implement. 
Do you think I should
a) Interpret `+ob` as `+o +b`
b) Disable the possibility to set several flags in the same command (interpret `+ob` as `+o`)

**Reply by Akuli** [reply]
I don't know. You can ask on #ircv3 on libera.

**Reply by ThePhilgrim** [reply]
Turns out this will be quite difficult to implement. 

I've discussed at length with some guys over at #ircv3.

Feel free to read (or not read) whatever parts of this discussion you want https://dpaste.com/FTA3L9MG8

**Comment by ThePhilgrim** [comment]
@Akuli Here is a discussion between me and val (progval) on #ircv3 regarding your latin-1 discussion:

https://dpaste.com/C43WRP2RS

**Reply by Akuli** [reply]
The discussion is now a bit scattered: there's some on irc, and some on the ircdocs issue. But everyone seem to agree that latin-1 works as long as you don't `.lower()` your strings, or do something else that requires you to interpret them as individual characters, rather than just bytes.

Indeed, the main purpose of the latin-1 trick is that we can use f-strings, if you want to put it that way :) We could change everything to `bytes`. It would work just the same, it wouldn't be too difficult, and `.lower()`ing user inputs would still work incorrectly.

**Comment by ThePhilgrim** [comment]
@Akuli 

Libera has quite a weird behavior when it comes to sending MODE messages.
For example:
`MODE #chan +o User` -> Sends msg to channel
`MODE #chan +o User` -> Sends msg to channel again
`MODE #chan +o User` -> Sends msg to channel again,
etc. (same with `-o`)

However:
`MODE #chan +b User` -> Sends msg to channel
`MODE #chan +b User` -> No msg to channel
`MODE #chan +b User` -> No msg to channel

`MODE #chan -b User` -> Sends msg to channel
`MODE #chan -b User` -> Sends msg to channel again
`MODE #chan -b User` -> Sends msg to channel again

I am considering not to adhere to this behaviour.
As it is now, for `+/-o`, I send the message every time.
However, for `+/-b` I send the message only once. In other words, if `+b` when user already is banned, I don't send the message again (same for `-b`)

What do you think is the best thing to do here? Follow Libera exactly, same behaviour for every flag, or like I'm doing now (consistency per flag, but different flags can differ)?

**Reply by Akuli** [reply]
I don't think sending the same message again solves a practical problem. If a user is already an operator, why should people be notified if someone mistakenly makes them an operator again? It could be confusing to users.

So my preference would be to not send messages about MODE changes that don't actually change anything.

**Reply by ThePhilgrim** [reply]
I definitely agree with this. I wonder why Libera is behaving differently.

**Comment by ThePhilgrim** [comment]
@Akuli I notice that some netcat tests in `test_mantatail.py` are using `run_server` fixture and some are not. Makes me wonder if the ones not using it are even working properly or if the one that are using it actually need it?

**Reply by Akuli** [reply]
Any test that uses `user_alice`, `user_bob` or `user_charlie` also automatically uses `run_server`, because the `user_blah` fixtures use `run_server`.

Speaking of tests, `test_mantatail.py` is getting quite long. We should split it up. It's common to put fixtures to `conftest.py` when you want the fixtures to be available in all test files.

**Reply by ThePhilgrim** [reply]
@Akuli But I also have a lot of tests using only `nc` that are not using `run_server` as fixture. For example:

```
def test_erroneus_nick():
    nc = socket.socket()
    nc.connect(("localhost", 6667))

    nc.sendall(b"NICK 123newnick\n")
    assert receive_line(nc) == b":mantatail 432 123newnick :Erroneous Nickname\r\n"

    nc.sendall(b"NICK /newnick\n")
    assert receive_line(nc) == b":mantatail 432 /newnick :Erroneous Nickname\r\n"

    nc.sendall(b"NICK newnick*\n")
    assert receive_line(nc) == b":mantatail 432 newnick* :Erroneous Nickname\r\n"
```

**Reply by Akuli** [reply]
Yeah, this test is bad. It will break if you run it by itself, e.g. `python3 -m pytest tests/test_mantatail.py::test_erroneous_nick`

**Reply by ThePhilgrim** [reply]
Interesting. How does it come then that when run together with the whole test suite, they pass?

**Reply by Akuli** [reply]
Nevermind, it will still work. I forgot that the `run_server` fixture is marked with `autouse=True`:

```python3
@pytest.fixture(autouse=True)
def run_server(fail_test_if_there_is_an_error_in_a_thread):
```
Because it has `autouse=True`, it will be used in every test, specified in the test or not.

**Comment by ThePhilgrim** [comment]
@Akuli 
Based on your comment about breaking up the test file, I'm thinking about doing some more splitting. 

The test file is currently 1200 lines.
I am not completely sure on how to split this. You mentioned `conftest.py` somewhere. What would this contain? And how do you think it makes sense to split up the rest? How do I use fixtures that are not in the same file as a test? I guess `import module; module.run_server; module.user_alice`. Correct?

`commands.py` is around 870 lines. I was planning to maybe break out the errors into `errors.py` (or similar), and maybe even the private functions? Do you think it makes sense to consider making the handler into a class or something? Or should I just keep the structure of the file as it is?

You also talked about splitting up `recv_loop()`since it's getting really long. I assume this means breaking out parts of it into separate functions. Would this then make sense to put in `class Receiver` (or similar), and make the different parts into methods?

**Reply by Akuli** [reply]
> You mentioned conftest.py somewhere. What would this contain? 

Fixtures.

> How do I use fixtures that are not in the same file as a test?

By putting the fixtures to `conftest.py`, you make them available to all `test_foo.py` files.

Let's look at Porcupine's tests, for example. Porcupine's conftest.py defines a `filetab` fixture, which is used in `test_tabs.py` and other `test_foo.py` files.

```python3
# tests/conftest.py
@pytest.fixture
def filetab(porcusession, tabmanager):
    tab = tabs.FileTab(tabmanager)
    tabmanager.add_tab(tab)
    return tab

# tests/test_tabs.py
def test_new_file_doesnt_show_up_as_modified(filetab):
    assert not filetab.has_unsaved_changes()
```

`receive_line()` and `compare_if_word_match_in_any_order()` are currently not fixtures, but apparently you must convert them into a fixture (or two separate fixtures) if you want to use them in multiple `test_foo.py` files. See https://stackoverflow.com/q/33508060, or if you prefer examples from Porcupine, `run_porcupine` and `wait_until` in Porcupine's conftest.py.

> And how do you think it makes sense to split up the rest? 

Looking at all tests we currently have with `grep 'def test_' tests/test_mantatail.py`, it seems that many of these are operator commands (op/deop, kick, etc). I would probably move those into a new file `test_operator_commands.py`, and see how much remains.

> I was planning to maybe break out the errors into errors.py (or similar)

I see the current error handling code as unnecessarily complex. For example, consider all the code involved in sending an error for erroneous nickname:

```python3
# commands.py
def handle_nick(...) -> None:
    ...
    error_erroneus_nickname(user, new_nick)
    ...


# commands.py (or maybe soon errors.py?)
def error_erroneus_nickname(user: mantatail.UserConnection, new_nick: str) -> None:
    (err_nick_num, err_nick_info) = irc_responses.ERR_ERRONEUSNICKNAME
    message = f"{err_nick_num} {new_nick} {err_nick_info}"
    user.send_que.put((message, "mantatail"))


# irc_responses.py
ERR_ERRONEUSNICKNAME = ("432", ":Erroneous Nickname")
```

I would probably write it like this:

```python3
# commands.py
def handle_nick(...) -> None:
    ...
    user.send_que.put((f"432 {new_nick} :Erroneous nickname", "mantatail"))
    ...
```

> splitting up `recv_loop` ... Would this then make sense to put in `class Receiver` (or similar)

If you aren't sure, try splitting it up with and without making a new class, and choose whichever code you prefer :)

**Comment by ThePhilgrim** [comment]
Hi @Akuli 

I was hoping you could help me. If you remember, a while back you were helping me to parse an XML file to make it more readable with `xml.dom.minidom`.

I am trying to use it again with another XML file, but I get the following error:

```
❯ python3 step1optional.py
Traceback (most recent call last):
  File "/Users/Writing/otherdev/xmlstripper/step1optional.py", line 4, in <module>
    dom = xml.dom.minidom.parse("ib2target.xml") # or xml.dom.minidom.parseString(xml_string)
  File "/usr/local/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/xml/dom/minidom.py", line 1988, in parse
    return expatbuilder.parse(file)
  File "/usr/local/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/xml/dom/expatbuilder.py", line 911, in parse
    result = builder.parseFile(fp)
  File "/usr/local/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/xml/dom/expatbuilder.py", line 207, in parseFile
    parser.Parse(buffer, False)
xml.parsers.expat.ExpatError: mismatched tag: line 1, column 96264
```

I was trying to find out if the "column" is the same as character # or similar. There seems to be invalid XML here, "mismatched tag"..
Would you know how I can figure out where in the file the problematic tag is? I really need to solve this as soon as possible since it's for work.
